doubleclick and longpress for PCF and MCP

giig1967g opened the pull request:
Added doubleclick and longpress events for PCF (P019) and MCP (P009).
Same logic as #1880.

giig1967g committed:
doubleclick and longpress for PCF and MCP

Added doubleclick and longpress events for PCF (P019) and MCP (P009)

giig1967g commented:
Added also invert logic for MCP.

TD-er commented:
I see code duplication between P001,  P019 and P009.
Is it possible to move this to a single function?
Or should we make a new issue to remove all this duplicate code?

giig1967g commented:
I will have a look into it to see if it is worth, as there could be a transfer of big structs between the functions

giig1967g committed:
fixed a display bug

giig1967g committed:
final fixes

TD-er commented on src/_P009_MCP.ino:
```diff
@@ -67,31 +124,151 @@ boolean Plugin_009(byte function, struct EventStruct *event, String& string)
+          Settings.TaskDevicePluginConfigFloat[event->TaskIndex][2] = PLUGIN_009_LONGPRESS_MIN_INTERVAL;
+
+        // @giig1967g-20181022: set initial UserVar of the switch
+        UserVar[event->BaseVarIndex] = switchstate[event->TaskIndex];
```
This was set in line 131 or 133.
Maybe set bootstate (line 137/138) before setting the Uservar values?

Current logic is too complicated to oversee all possibilities of pin state inverse and boot state.

giig1967g replied:
will review it

giig1967g replied:
> This was set in line 131 or 133.

Right. Fixed.

> Maybe set bootstate (line 137/138) before setting the Uservar values?

Will do that. But it won't matter, as immidiately after, PLUGIN_TEN_PER_SECOND will overwrite switchstate and UserVar anyway.

> Current logic is too complicated to oversee all possibilities of pin state inverse and boot state.

I have reviewed the whole logic and will post it soon

TD-er commented on src/_P009_MCP.ino:
```diff
@@ -67,31 +124,151 @@ boolean Plugin_009(byte function, struct EventStruct *event, String& string)
             sendData(event);
+
+            //reset Userdata so it displays the correct state value in the web page
+            UserVar[event->BaseVarIndex] = sendState ? 1 : 0;
```
I get it why you want to reset the UserVar value.
I don't know if the event in `sendData` is being prepared fully before exit. (collecting all values for further processing to controllers)
Maybe worth to check if it is indeed being processed (stored in the controller delay queue) or else it will be very hard to debug errors later.

giig1967g replied:
Good point. But I need your help in how to check if sendData has finished or not... ;)

TD-er commented on src/_P019_PCF8574.ino:
```diff
@@ -40,7 +40,7 @@ TaskDevicePluginConfigLong settings:
 {
   boolean success = false;
-  static int8_t switchstate[TASKS_MAX];
+  static int switchstate[TASKS_MAX];
```
Why changing it to an int?

The build error is about this line in both files (P009 and P019)
`if (state != -1 && state != switchstate[event->TaskIndex])`
`state` is a boolean, not an int and thus will this check always return false.

giig1967g replied:
I was trying to solve a cosmetic issue on displaying the status of non connected port.
But after digging I found that current implementation has a bug: 
Assume a user creates a device and selects a port of PCF or MCP that doesn't exist or is not connected. The internal state of that port is -1, but with current implementation it's converted to 0. I was trying to fix the cosmetic thus I changed the data type.

But I will fix the internal logic too.
In the case presented above I will display -1 so user will now that the port is incorrect or disconnected or simply is not found.

giig1967g replied:
In my environement it was compiling fine. Strange.

TD-er replied:
The Travis settings are to consider some warnings as error.
And this is certainly one to consider as an error, since it is unexpected behavior.

giig1967g committed:
fixed as requested

- Fixed bootstate logic and inversed logic
- added selected events for doubleclick and longpress
- fixed log messages
- cleaned code

giig1967g commented:
New version:
- Added debounce for PCF and MCP
- Fixed bootstate logic and inversed logic
- Fixed variable types warnings
- Events for doubleclick and longpress are now selectable
- Fixed log messages
- Restored state = -1 for ports not available (PCF and MCP)
- Cleaned code
- Commented all the changes in the code

giig1967g reviewed (commented):
please check new version

TD-er commented on src/_P001_Switch.ino:
```diff
@@ -324,7 +364,10 @@ boolean Plugin_001(byte function, struct EventStruct *event, String& string)
+              if (Settings.TaskDevicePin1Inversed[event->TaskIndex])
+                sendState = !sendState;
+
+              if (Settings.TaskDevicePluginConfig[event->TaskIndex][7]==3 && Settings.TaskDevicePluginConfig[event->TaskIndex][4]>0)
```
Those `COUNTER` and `DC` defines may be extended to here also (maybe the whole plugin?)
And to avoid future redefine issues, maybe prefix the name with the plugin number?

It makes the code a lot more readable and maybe we should do it on more plugins.

giig1967g replied:
The idea of DC and COUNTER were only to simplify the "if" clause.
I could change them to P001_DC and P001_COUNTER and make the change in all 3 plugin for all the settings.
In order to schedule the work, is this the only change requested?

TD-er replied:
> In order to schedule the work, is this the only change requested?

 It was more like a comment, not really a change request.
I like the extra readability it gives.

giig1967g replied:
OK. Considering that the Switch, PCF and MCP plugin will be rewritten eventually, could we leave the  COUNTER and DC defines for next time?
So this PR is mergeable?

TD-er replied:
Yep, if nobody objects, I will merge it. :)
